[ui] Ask user before closing expression dialog if expression was edited

kannes opened the pull request:
## Description
The expression dialog makes it easy to lose one's work if the user hits the Esc key and closes the dialog. Any edits since the last press on OK will be lost.

I added a simple check in the dialog that compares the text that was in it when opened and the text that is in it when it is supposed to be closed without applying (rejecting with Cancel/Esc/X, not accepting the dialog with OK). If there is a different, a QMessageBox will ask the user if they really want to throw their changes away.

The user can choose to not get this warning again by ticking a checkbox, similar to other dialogs. The choice is saved as settings option `gui/askToDiscardEditedExpression`.

![grafik](https://user-images.githubusercontent.com/7661092/187282261-702495c1-1c2c-4353-92ba-54e8cd9b374b.png)

With this change it affects the "Expression String Builder" (Override icon -> edit), "Expression Dialog (Epsilon icon, e. g. for Geometry Generator or for Pre-calculated Value in the model builder), "Expression Based Filter" (Attribute Table -> Advanced filter expression). I am not sure if there are any other Epsilon button dialogs that are different and not affected by the change.

It does not affect "Select by Expression", "Raster Calculator", "Mesh Calculator" and other potential uses of the dialog. If wanted I could add the same checks to those dialogs.

My original motivation to add a safe-guard was when discarding a autocomplete suggestion with Esc did close the dialog. I could not reproduce that anymore though. Still I think this would be a nice feature to have.

I [posted it on Twitter](https://twitter.com/cartocalypse/status/1561361521887903745) and got some mixed feedback. A frequent feedback was to have an option to disable it. Several people stated that they lost expressions due to this in the past and would really like a safeguard like this.

I like [Matthias' suggestion](https://twitter.com/mprove/status/1561374771195019265) to have an Apply button and automatically saving edit state but that would be way beyond my non-existent C++ skills. Another feedback IRL was that a undo that survive closing the dialog would be great. And another idea was to have a list of stored expressions available similar to the functions editor for the Python functions.

# TODO
## Add to other dialogs?
Should this be added to "Select by Expression", "Raster Calculator", "Mesh Calculator" and potentially other dialogs as well?

## BadWindow warnings!
Each time the QMessageBox is triggered, I get a BadWindow warning in the terminal. Here are some from earlier:

```
Warning: QXcbConnection: XCB error: 3 (BadWindow), sequence: 3320, resource id: 21713376, major code: 40 (TranslateCoords), minor code: 0
Warning: QXcbConnection: XCB error: 3 (BadWindow), sequence: 4674, resource id: 21717126, major code: 40 (TranslateCoords), minor code: 0
Warning: QXcbConnection: XCB error: 3 (BadWindow), sequence: 5052, resource id: 21720427, major code: 40 (TranslateCoords), minor code: 0
Warning: QXcbConnection: XCB error: 3 (BadWindow), sequence: 5263, resource id: 21722496, major code: 40 (TranslateCoords), minor code: 0
```

How can I avoid them? 

# Warning or Question as QMessageBox type?
Not sure. I think Question fits better.

# Which options directory to put the option in?
`QgsSettings::App`? `QgsSettings::Gui`?


## Documentation?
Seems not needed to me.

## Tests?
I would need some help if tests are needed for this.

# Tested
- [x] I compiled and ran QGIS with this change and it worked fine for me.

-----


Greetings from the contributor's meeting in Firenze and thanks to the lovely people around for their help!

kannes committed:
Ask user before closing expression dialog if expression was edited

kannes committed:
Fix formatting via astyle

kannes commented:
Suggestion via @uclaros that I also had in mind: Put the focus on "No" so getting out of the editor is just "Esc -> Enter", not "Esc -> Tab -> Tab -> Tab confusedly until user is really sure focus is on the right button -> Enter" :)

That would be another `QMessageBox::Yes` passed to the `QMessageBox::warning()`.

nyalldawson commented:
I'm a big +1 to this. I've lost many an expression from accidentally hitting ESC in an misguided attempt to close the autocomplete pop ups...

kannes commented:
Yay! :)
I will think some more about the feedback I got so far and will probably try to make it only trigger when the ESC key is involved (so a keyPressEvent filter rather than reject() I guess). That's seems like the main (or sole) motivation for it and I think then everyone could agree that we wouldn't need an option to "don't ask me again" (which would feel like total overkill to add to the options menu...). My plan is to finalize this during FOSS4G or the code sprint on saturday.

nirvn commented:
@kannes , feel like implementing the [x] Show this again next time checkbox ? :)

kannes commented:
Guess what I am doing today by copying from your https://github.com/qgis/QGIS/pull/49977 ;)

kannes committed:
Add option to disable warning about losing changes

- Users can disable the warning
- Adds a settings option in gui/askToDiscardEditedExpression (changeable via Advanced options)
- Wording similar to other existing dialogs

Thanks to @nirvn as I re-based this on his https://github.com/qgis/QGIS/pull/49977

nyalldawson commented on src/gui/qgsexpressionbuilderdialog.cpp:
```diff
@@ -83,6 +86,40 @@ void QgsExpressionBuilderDialog::accept()
+    if ( askToDiscardEditedExpression )
+    {
+      QMessageBox confirmMessage( QMessageBox::Question,
+                                  tr( "Expression was edited" ),
```
```suggestion
                                  tr( "Expression was Edited" ),
```

kannes replied:
Thanks for looking at it :) Why should that word be uppercase?

kannes replied:
Ah, because of https://docs.qgis.org/testing/en/docs/developers_guide/hig.html :)

nyalldawson commented:
Lgtm!

You'll need to run the scripts/sipify_all.sh script to update the sip bindings and get the tests all passing

uclaros reviewed (commented):
`mInitialText` can be const
Also, have a look at https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md on how you can automate the formatting and sipifying steps on all commits!

kannes commented on src/gui/qgsexpressionbuilderdialog.cpp:
```diff
@@ -17,6 +17,7 @@
+#include <QMessageBox>
 
 QgsExpressionBuilderDialog::QgsExpressionBuilderDialog( QgsVectorLayer *layer, const QString &startText, QWidget *parent, const QString &key, const QgsExpressionContext &context )
   : QDialog( parent )
```
Oh nice, I didn't know one could do that.

kannes committed:
Make mInitialText a const

Via review by @uclaros, thanks!

Co-authored-by: Stefanos Natsis <uclaros@gmail.com>

kannes committed:
Use titlecase for dialog title

Co-authored-by: Nyall Dawson <nyall.dawson@gmail.com>

kannes committed:
Adjust wording and button text

kannes committed:
Add new reject method in .sip.in

kannes commented:
I added the suggested changes, the SIP stuff and also adjusted the wording and GUI further to match @nirvn's recent https://github.com/qgis/QGIS/pull/49977 because consistency in the UI is super important to me.

Ready to be merged from my point of view. As of right now the 22.04 test has not finished but the others were all fine.

lbartoletti commented:
@kannes you have to generate the sip files. You could use `scripts/sipify_all.sh`

lbartoletti commented:
And BTW, we have pre-commit hook: https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md

kannes commented:
The web interface hid the fact that there was more than 6 checks, sorry.

To be honest I am juggling with the Github web interface for the commits right now because I could not be arsed to set up all the kerfuffle required to push nowadays on my holiday system. :\ I did run the script(s) and checked the diff but apparently I missed a single additional newline in the .sip.in. Let' see...

kannes committed:
Add missing linebreak?

uclaros commented:
Nicely done @kannes !

kannes commented:
Thanks a lot for the help on my baby steps and the discussion you all, this makes me super happy!
